File: assign4/javascript/test/guess_game.test.js
'use strict';

const { guess,  playGame, getRandomNumber, GuessResult: { LOW, CORRECT, HIGH } } = require('../src/guess_game');

describe('canary test', () => {
  it('should pass a canary test', () => {
    expect(true).toBe(true);
  });
});

describe('guess function', () => {
  it('should return "low" when the guess is less than the target', () => {
    const result = guess(10, 5);

    expect(result).toBe(LOW);
  });

  it('should return "high" when the guess is more than the target', () => {
    const result = guess(10, 15);

    expect(result).toBe(HIGH);
  });

  it('should return "correct" when the guess is the target', () => {
    const result = guess(10, 10);

    expect(result).toBe(CORRECT);
  });
});

describe('playGame takes a target, a function that returns the guess, a function that will display the result', () => {
  it('win on the first attempt', () => {
    const target = 10;
    const messages = [];

    const readGuess = () => target;
    
    const display = (message) => { messages.push(message); };

    const attempts = playGame(target, readGuess, display);

    expect(attempts).toBe(1);
    expect(messages).toEqual([CORRECT]);
  });

  it('win on the second attempt', () => {
    const target = 10;
    const guesses = [3, 10];
    const messages = [];

    const readGuess = () => guesses.shift();
    const display = (message) => { messages.push(message); };

    const attempts = playGame(target, readGuess, display);

    expect(attempts).toBe(2);
    expect(messages).toEqual([LOW, CORRECT]);
  });
  
  it('win on the third attempt', () => {
    const target = 10;
    const guesses = [3, 12, 10];
    const messages = [];

    const readGuess = () => guesses.shift();
    const display = (message) => { messages.push(message); };

    const attempts = playGame(target, readGuess, display);

    expect(attempts).toBe(3);
    expect(messages).toEqual([LOW, HIGH, CORRECT]);
  });
});

describe('get random number', () => {
  it('number is between 1 and 100, inclusive', () => {
    const target = getRandomNumber();

    expect(1 <= target <= 100).toBeTruthy();
  });

  it('number is random', () => {
    const seed1 = 1;
    const target1 = getRandomNumber(seed1);

    const seed2 = 2;
    const target2 = getRandomNumber(seed2);

    expect(target1 !== target2).toBeTruthy();
  });
});

File: assign4/f#/test/guess_game.test.fs
namespace test

open System
open NUnit.Framework
open src.GuessGame

[<TestFixture>]
type GuessGameTest () =

  [<Test>]
  member this.CanaryTest() =
    Assert.That(true, Is.True)

  [<Test>]
  member this.GuessFunctionLOW() =
    let result = guess 10 5

    Assert.That(result, Is.EqualTo(GuessResult.LOW))
        
  [<Test>]
  member this.GuessFunctionHIGH() =
    let result = guess 10 15

    Assert.That(result, Is.EqualTo(GuessResult.HIGH))

  [<Test>]
  member this.GuessFunctionCORRECT() =
    let result = guess 10 10

    Assert.That(result, Is.EqualTo(GuessResult.CORRECT))

  [<Test>]
  member this.playGameWinOnFirstAttempt() =
    let target = 10
    let messages = ResizeArray<GuessResult>() 

    let inline readGuess _ = 10
    let inline display message = messages.Add message

    let attempts = playGame(target, readGuess, display)

    Assert.That(messages.ToArray(), Is.EqualTo([|GuessResult.CORRECT|]))

  [<Test>]
  member this.playGameWinOnSecondAttempt() =
    let target = 10
    let mutable inputs = [|5; 10|]
    let messages = ResizeArray<GuessResult>() 

    let inline readGuess _ = 
      let guess = inputs.[0]
      inputs <- inputs.[1..]
      guess

    let inline display message = messages.Add message

    let attempts = playGame(target, readGuess, display)

    Assert.That(messages.ToArray(), Is.EqualTo([|GuessResult.LOW; GuessResult.CORRECT|]))

  [<Test>]
  member this.playGameWinOnThirdAttempt() =
    let target = 10
    let mutable inputs = [|5; 15; 10|]
    let messages = ResizeArray<GuessResult>() 

    let inline readGuess _ = 
      let guess = inputs.[0]
      inputs <- inputs.[1..]
      guess

    let inline display message = messages.Add message

    let attempts = playGame(target, readGuess, display)

    Assert.That(messages.ToArray(), Is.EqualTo([|GuessResult.LOW; GuessResult.HIGH; GuessResult.CORRECT|]))

  [<Test>]
  member this.numberBetween1And100() =
    let seed = 42
    let target = getRandomNumber(seed) 

    Assert.That(target, Is.GreaterThanOrEqualTo(1).And.LessThanOrEqualTo(100))

  [<Test>]
  member this.numberIsRandom() =
    let seed1 = 1
    let target1 = getRandomNumber(seed1)

    let seed2 = 2
    let target2 = getRandomNumber(seed2)

    Assert.That(target1, Is.Not.EqualTo(target2))
